aten/src/ATen/core/ITensorList_test.cpp

```diff
@@ -0,0 +1,166 @@
+#include <ATen/Functions.h>
+#include <ATen/core/ITensorList.h>
+#include <gtest/gtest.h>
+
+using namespace c10;
+
+static std::vector<at::Tensor> get_tensor_vector() {
+  std::vector<at::Tensor> boxed;
+  const size_t SIZE = 5;
+  for (size_t i = 0; i < SIZE; i++) {
+    boxed.push_back(at::empty({0}));
+  }
+  return boxed;
+}
+
+template <typename T>
+void check_elements_same(ITensorList list, const T& thing) {
+  EXPECT_EQ(thing.size(), list.size());
+  for (size_t i = 0; i < thing.size(); i++) {
+    EXPECT_TRUE(thing[i].is_same(list[i]));
+  }
+}
+
+TEST(ITensorListTest, CtorEmpty_IsNone_Throws) {
+  ITensorList list;
+  EXPECT_TRUE(list.isNone());
+  // NOLINTNEXTLINE(cppcoreguidelines-avoid-goto,hicpp-avoid-goto)
+  EXPECT_THROW(list.size(), c10::Error);
+}
+
+TEST(ITensorListTest, CtorBoxed_IsBoxed) {
+  auto vec = get_tensor_vector();
+  List<at::Tensor> boxed(vec);
+  ITensorList list(boxed);
+  EXPECT_TRUE(list.isBoxed());
+}
+
+TEST(ITensorListTest, CtorUnboxed_IsUnboxed) {
+  auto vec = get_tensor_vector();
+  at::ArrayRef<at::Tensor> unboxed(vec);
+  ITensorList list(unboxed);
+  EXPECT_TRUE(list.isUnboxed());
+}
+
+TEST(ITensorListTest, CtorUnboxedIndirect_IsUnboxed) {
+  auto vec = get_tensor_vector();
+  auto check_is_unboxed = [](ITensorList list) {
+    EXPECT_TRUE(list.isUnboxed());
+  };
+  check_is_unboxed(vec[0]);
+  check_is_unboxed({vec.data(), vec.size()});
+  check_is_unboxed({&*vec.begin(), &*vec.end()});
+  check_is_unboxed(vec);
+}
+
+TEST(ITensorListTest, CtorTemp_IsUnboxed) {
+  auto check_is_unboxed = [](ITensorList list) {
+    EXPECT_TRUE(list.isUnboxed());
+  };
+
+  auto vec = get_tensor_vector();
+  check_is_unboxed({vec[0], vec[1]});
+}
+
+TEST(ITensorListTest, Boxed_GetConstRefTensor) {
+  auto vec = get_tensor_vector();
+  // Using 'const' here, so that 'operator[]' returns
+  // a 'Tensor' reference (plus, we don't need to mutate it here).
+  const List<at::Tensor> boxed(vec);
```

Comment:
It seems to me like this should work even if List<at::Tensor> is not const? Don't quite understand the comment here.

Reply:
You are right. I think this is a stale test case. I think I was using `boxed_type*` instead of `const boxed_type*` in the union.

Reply:
Ah. So, there was a reason for this `const`! The problem is that `List::operator[]` returns a `ListElementReference` instead of returning a `Tensor`. On the other hand, `List::operator[] const` returns a `const Tensor&`.

Reply:
Gotcha. This is kind of goofy but an explainer is probably enough, you don't have to make up for List's problems.